libtiledbvcf/src/utils/logger.cc

```diff
@@ -0,0 +1,198 @@
+#include <spdlog/sinks/basic_file_sink.h>
+#include <spdlog/sinks/stdout_color_sinks.h>
+
+namespace tiledb::common {
+
+// Set the default logging format
+// %^ : start color range
+// [Year-month-day 24hr-min-second.microsecond]
+// [logger]
+// [Process: id]
+// [Thread: id]
+// [log level]
+// text to log...
+// %$ : end color range
+const std::string LOG_PATTERN =
+    "%^[%Y-%m-%d %H:%M:%S.%e] [%n] [Process: %P] [Thread: %t] [%l] %v%$";
+const std::string CONSOLE_LOGGER = "tiledb-vcf";
+const std::string FILE_LOGGER = "tiledb-vcf-file";
+
+/* ********************************* */
+/*     CONSTRUCTORS & DESTRUCTORS    */
+/* ********************************* */
+
+Logger::Logger() {
+  logger_ = spdlog::get(CONSOLE_LOGGER);
+  if (logger_ == nullptr) {
+    logger_ = spdlog::stdout_color_mt(CONSOLE_LOGGER);
+  }
+  logger_->set_pattern(LOG_PATTERN);
+  set_level("FATAL");
+}
+
+Logger::~Logger() {
+  spdlog::drop(CONSOLE_LOGGER);
+  if (spdlog::get(FILE_LOGGER) != nullptr) {
+    spdlog::drop(FILE_LOGGER);
+  }
+}
+
+void Logger::trace(const char* msg) {
+  logger_->trace(msg);
+}
+
+void Logger::debug(const char* msg) {
+  logger_->debug(msg);
+}
+
+void Logger::info(const char* msg) {
+  logger_->info(msg);
+}
+
+void Logger::warn(const char* msg) {
+  logger_->warn(msg);
+}
+
+void Logger::error(const char* msg) {
+  logger_->error(msg);
+}
+
+void Logger::critical(const char* msg) {
+  logger_->critical(msg);
+}
+
+void Logger::set_level(const std::string& level) {
+  if (level == "FATAL") {
+    level_ = spdlog::level::critical;
+  } else if (level == "ERROR") {
+    level_ = spdlog::level::err;
+  } else if (level == "WARN") {
+    level_ = spdlog::level::warn;
+  } else if (level == "INFO") {
+    level_ = spdlog::level::info;
+  } else if (level == "DEBUG") {
+    level_ = spdlog::level::debug;
+  } else if (level == "TRACE") {
+    level_ = spdlog::level::trace;
+  } else {
+    set_level("WARN");
+    LOG_WARN("Illegal log level = {}, using log level FATAL", level);
+    level_ = spdlog::level::critical;
+  }
+  logger_->set_level(level_);
+}
+
+void Logger::set_logfile(const std::string& filename) {
+  try {
+    auto file_logger = spdlog::basic_logger_mt(FILE_LOGGER, filename);
+    file_logger->set_pattern(LOG_PATTERN);
+    file_logger->set_level(level_);
+    file_logger->flush_on(spdlog::level::warn);
+  } catch (spdlog::spdlog_ex& e) {
+    // log message and exit if file logger cannot be created
+    LOG_FATAL(e.what());
+  }
+
+  // add sink to existing logger
+  // (https://github.com/gabime/spdlog/wiki/4.-Sinks)
+  auto file_sink = spdlog::get(FILE_LOGGER)->sinks().back();
+  logger_->sinks().push_back(file_sink);
+}
+
+bool Logger::debug_enabled() {
+  return (level_ == spdlog::level::debug) || (level_ == spdlog::level::trace);
+}
+
+/* ********************************* */
+/*              GLOBAL               */
+/* ********************************* */
+
+Logger& global_logger() {
+  static Logger l;
```

Comment:
Instead of having a single global logger at the process level could we have loggers set on the dataset/reader/writer level or similar? With spark usage it'd be nice to not have a single logger but have separate ones were we eventually have tags such as partition ids etc so its clear which logs go to which query.

Reply:
This doesn't have to be done in this initial version, just a though.

Reply:
Yes, that makes sense. Maybe this PR can cover CLI logging and we can target multiple loggers in the next version.

Reply:
Sounds good